src/math/matrix.c

```diff
@@ -3,93 +3,112 @@
 // Copyright(c) 2022 Intel Corporation. All rights reserved.
 //
 // Author: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
+//	   Shriram Shastry <malladi.sastry@linux.intel.com>
+//
 
 #include <sof/math/matrix.h>
 #include <errno.h>
 #include <stdint.h>
 
-int mat_multiply(struct mat_matrix_16b *a, struct mat_matrix_16b *b, struct mat_matrix_16b *c)
+/* Performs matrix multiplication of two fixed-point 16-bit integer matrices,
+ * storing the result in a third matrix. It accounts for fractional bits for
+ * fixed-point arithmetic, adjusting the result accordingly.
+ *
+ * Arguments:
+ *   a: pointer to the first input matrix
+ *   b: pointer to the second input matrix
+ *   c: pointer to the output matrix to store result
+ *
+ * Return:
+ *   0 on successful multiplication.
+ *   -EINVAL if input dimensions do not allow for multiplication.
+ *   -ERANGE if the shift operation might cause integer overflow.
+ */
+int mat_multiply(struct mat_matrix_16b *a, struct mat_matrix_16b *b,
+		 struct mat_matrix_16b *c)
 {
-	int64_t s;
-	int16_t *x;
-	int16_t *y;
-	int16_t *z = c->data;
-	int i, j, k;
-	int y_inc = b->columns;
-	const int shift_minus_one = a->fractions + b->fractions - c->fractions - 1;
+	int32_t acc;	/* Accumulator for dot product calculation */
+	int16_t *x, *y, *z = c->data; /* Pointers for matrices a, b, and c */
+	int i, j, k;	/* Loop counters */
+	int y_inc = b->columns;	   /* Column increment for matrix b elements */
+	/* Calculate shift amount for adjusting fractional bits in the result */
+	const int shift = a->fractions + b->fractions - c->fractions - 1;
```

Comment:
is c not fully output? I.e. are some fields expected to be set beforehand? Can we get docs explaining that?

Reply:
I 'm not sure on documentation but yeah - `mat_multiply` function and the initialization state of the c matrix is described below. 

Before calling `mat_multiply`, the caller is expected to allocate the c matrix using `mat_matrix_alloc_16b`, specifying the dimensions and fractional bits. This means the `c` matrix will already have appropriate memory allocated for its data, and its `rows`, `columns`, and `fractions` fields will be initialized accordingly. However, the function `mat_multiply` itself does not modify these metadata fields of the `c` matrix; instead, it assumes that the `c` matrix's dimensions are correctly set to accommodate the result of the matrix multiplication (i.e., `c->rows` should be equal to `a->rows` and `c->columns` should be equal to `b->columns` for standard matrix multiplication).

`mat_multiply` calculates the product of matrices `a` and `b`, and stores the result in `c`. The function internally calculates the necessary adjustment for fractional bits based on the fractional bits of the input matrices (`a` and `b`) and the output matrix (`c`). It ensure that the result stored in `c` conforms to the expected fixed-point representation.